### PR TITLE
0D, 1D and zero volume support for generic reduce ops

### LINKED ISSUE #16623: 0D/1D support for generic reduce
From @ayerofieiev-tt 

There is probably also a bug for 1D tensors in ops like Sum
If user specifies a dimension over which to make a reduction and an input tensor is 1D...
0 dim will actually mean 1 dim

Will need to look into this once 0D/1D tiled tensor support is fully in.

Would also be good to look at 0 volume tensors (one dimension has size 0). E.g. for 2D and 4D.

### PR DESCRIPTION
**Pull request by nsorabaTT**
### Ticket
[Link to Github Issue](https://github.com/tenstorrent/tt-metal/issues/16623)

### Problem description
Currently, generic reduction operations only support non-zero volume tensors and tensors with rank >=2.

[This](https://github.com/tenstorrent/tt-metal/issues/19480) and [this](https://github.com/tenstorrent/tt-metal/issues/19478) impact 1D std/var and min/max as well. To be taken up separately.

### What's changed
This PR adds the following:

1. Support 0D and 1D tensors for all generic ops
2. Support zero volume tensors of any rank (0-4)
4. Add a sweep to test all generic reduction ops across all zero and non zero volume tensors for ranks 0-4.
5. Add unit tests for handpicked cases from above
6. Allow "scalar" to propagate to all generic reduce ops.

| Reduce Op | All ranks supported  | Zero volume supported? | Matches Torch?  | 
|-----------|----------------------|------------------------|-----------------|
| `sum`     |          Y           |            Y           |        Y        |
| `mean`    |          Y           |            Y           |        Y        |
| `max`     |          Y           |            Y           |        Y        |
| `min`     |          Y           |            Y           |        Y        |
| `std`     |          Y           |            Y           |        N        |
| `var`     |          Y           |            Y           |        N        |

#### Review hint:
`f3f2be19bb8349a530d6263aa7d44a012e907ad4` cleans up existing sweeps.
Rest adds new features.

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) [CI 28265 passing ](https://github.com/tenstorrent/tt-metal/actions/runs/14041777945)
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] **(For models and ops writers)** Full [new models tests](https://github.com/tenstorrent/tt-metal/actions/workflows/full-new-models-suite.yaml) CI passes (if applicable)
- [x] New/Existing tests provide coverage for changes
- [x] [ttnn - Run sweeps #2457](https://github.com/tenstorrent/tt-metal/actions/runs/14041766008)

### CODE REVIEW

File: tests/ttnn/unit_tests/operations/reduce/test_reduction.py [hunk 1]
```diff
@@ -412,3 +412,68 @@ def run_reduce_sum_h(device, batch_size, h, w, dim):
...
+        torch_result.values
+        if isinstance(torch_result, (torch.return_types.min, torch.return_types.max))
+        else torch_result
+    )
+
+    ttnn_result = ttnn.to_torch(ttnn.from_device(ttnn_result))
+
+    assert_with_pcc(torch_result, ttnn_result, 0.99)
+
+    if op not in ["std", "var"]:
```
Comment (bbradelTT) [inline tests/ttnn/unit_tests/operations/reduce/test_reduction.py:474]: Can we check allclose for all ops? Would atol or rtol  need to change? That will help with regressions.

Comment (nsorabaTT) [reply tests/ttnn/unit_tests/operations/reduce/test_reduction.py:474]: Yes. I tested the following change and it works:

```python
atol = rtol = 0.1
if op == "std":
        atol, rtol = sys.maxsize, 0.1 + math.sqrt(2)
elif op == "var":
        atol, rtol = sys.maxsize, 0.1 + 2
```
Then we can test allclose for all ops. Does this look good?

File: tests/sweep_framework/sweeps/reduction/generality/all_ranks.py
Comment (jdesousa-TT) [inline tests/sweep_framework/sweeps/reduction/generality/all_ranks.py:1]: This file does not describe a sweep according to the sweep framework. See README here https://github.com/tenstorrent/tt-metal/blob/e725458329c4205d71fd7d8252f61122cce33bbf/tests/sweep_framework/README.md

**Review by vsureshTT — APPROVED** [review]

File: tests/ttnn/unit_tests/operations/reduce/test_reduction.py (hunk 1 above)
Comment (nsorabaTT) [reply tests/ttnn/unit_tests/operations/reduce/test_reduction.py]: Addressed in `bba7268aeb0131a90606d42dd24f72d1b225c08e`

File: tests/sweep_framework/sweeps/matmul/full/matmul_default_block_sharded.py [hunk 2]
```diff
@@ -173,51 +173,27 @@ def test_matmul(
     input_layout,
     compute_kernel_config,
 ):
-    run_matmul(
+    run_reduction(
```
Comment (jdesousa-TT) [inline tests/sweep_framework/sweeps/matmul/full/matmul_default_block_sharded.py:176]: Intentional change?

File: tests/sweep_framework/sweeps/reduction/generality/all_ranks.py
Comment (jdesousa-TT) [reply tests/sweep_framework/sweeps/reduction/generality/all_ranks.py:1]: Please run this new sweep in the sweeps pipeline before merging. https://github.com/tenstorrent/tt-metal/actions/workflows/ttnn-run-sweeps.yaml

File: tests/sweep_framework/sweeps/reduction/generality/reduction_all_ranks.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+        atol, rtol = sys.maxsize, 0.1 + 2
+
+    return [
+        torch.allclose(torch_result, ttnn_result, atol=atol, rtol=rtol, equal_nan=True),
+        f"mismatch in allclose: torch: {torch_result}, ttnn: {ttnn_result}",
+    ]
+
+
+@pytest.mark.parametrize(**gen_pytest_parametrize_args(parameters))
+def test_matmul(
```
Comment (bbradelTT) [inline tests/sweep_framework/sweeps/reduction/generality/reduction_all_ranks.py:105]: test_reduction maybe?

**Review by bbradelTT — APPROVED** [review]
Looks good.

Please
- change the test_matmul name
- run the sweeps test via pytest
- run the sweeps test via the sweeps framework
- have APC pass

File: tests/sweep_framework/sweeps/reduction/generality/reduction_all_ranks.py (hunk 3 above)
Comment (nsorabaTT) [reply tests/sweep_framework/sweeps/reduction/generality/reduction_all_ranks.py]: Fixed in `174d491a70aed172531e044e972e755c245c783b`

File: tests/sweep_framework/sweeps/matmul/full/matmul_default_block_sharded.py (hunk 2 above)
Comment (nsorabaTT) [reply tests/sweep_framework/sweeps/matmul/full/matmul_default_block_sharded.py]: Fixed in `84651526f4c7faab977ed87efce31d87611cd6a8`

File: tests/sweep_framework/sweeps/reduction/generality/all_ranks.py
Comment (nsorabaTT) [reply tests/sweep_framework/sweeps/reduction/generality/all_ranks.py:1]: Ran sweeps and it passed. Update PR description with the link as well.

[ttnn - Run sweeps #2457](https://github.com/tenstorrent/tt-metal/actions/runs/14041766008)

**Review by jdesousa-TT — APPROVED** [review]
